src/modules/utils/motordrivercontrol/MotorDriverControl.cpp

```diff
@@ -112,34 +110,79 @@ bool MotorDriverControl::config_module(uint16_t cs)
         chip= TMC2660;
         tmc26x= new TMC26X(std::bind( &MotorDriverControl::sendSPI, this, _1, _2, _3), axis);
 
+    }else if(str == "TMC2208") {
+        chip= TMC2208;
+        tmc22x= new TMC22X(std::bind( &MotorDriverControl::sendUART, this, _1, _2, _3), axis);
+
     }else{
         THEKERNEL->streams->printf("MotorDriverControl %c ERROR: Unknown chip type: %s\n", axis, str.c_str());
         return false;
     }
 
-    // select which SPI channel to use
-    int spi_channel = THEKERNEL->config->value(motor_driver_control_checksum, cs, spi_channel_checksum)->by_default(1)->as_number();
-    int spi_frequency = THEKERNEL->config->value(motor_driver_control_checksum, cs, spi_frequency_checksum)->by_default(1000000)->as_number();
+    if(chip == TMC2208) {
+        //Configure soft UART
+    		
+        //select TX and RX pins
+        sw_uart_tx_pin = new Pin();
+        sw_uart_rx_pin = new Pin();
+        sw_uart_tx_pin->from_string(THEKERNEL->config->value(motor_driver_control_checksum, cs, sw_uart_tx_pin_checksum)->by_default("nc")->as_string())->as_output();
+        sw_uart_rx_pin->from_string(THEKERNEL->config->value(motor_driver_control_checksum, cs, sw_uart_rx_pin_checksum)->by_default("nc")->as_string())->as_input();
+
+        if(!sw_uart_tx_pin->connected() || !sw_uart_rx_pin->connected()) {
+            if(!sw_uart_tx_pin->connected()) {
+                THEKERNEL->streams->printf("MotorDriverControl %c ERROR: uart tx pin not defined\n", axis);
+            }
+            if(!sw_uart_rx_pin->connected()) {
+                THEKERNEL->streams->printf("MotorDriverControl %c ERROR: uart rx pin not defined\n", axis);
+            }
+            return false; // if not defined then we can't use this instance
+        }
+
+        //select soft UART baudrate
+        int sw_uart_baudrate = THEKERNEL->config->value(motor_driver_control_checksum, cs, sw_uart_baudrate_checksum)->by_default(9600)->as_number();
 
-    // select SPI channel to use
-    PinName mosi, miso, sclk;
-    if(spi_channel == 0) {
-        mosi = P0_18; miso = P0_17; sclk = P0_15;
-    } else if(spi_channel == 1) {
-        mosi = P0_9; miso = P0_8; sclk = P0_7;
+        PinName txd = port_pin((PortName)(sw_uart_tx_pin->port_number<<8), sw_uart_tx_pin->pin);
+        PinName rxd = port_pin((PortName)(sw_uart_rx_pin->port_number<<8), sw_uart_rx_pin->pin);
+
+        this->serial = new BufferedSoftSerial(txd, rxd);
```

Comment:
As the Pin is not used after this and the BufferedSerial uses MBED pinnames and not the Pin class, why not delete the Pins here as they are never used again. Then move the BufferedSerial into the union instead of the Pin* to save memory?

Reply:
Also maybe have an option to only use the TX pin and then disable all read functions, as that appears to be a recommended use case for this chip (write only). Most people will not have enough memory available anyway to have more than 3 drivers when each has a buffered serial for each pin.